fix(core): implement ldml_processor::get_key_list() 🙀

srl295 opened the pull request:
- ldml::vkeys class updated to keep a set<> of keys for the get_key_list() implementation
- note that ALL keys are included, not just assigned keys.
- all EmbeddedLdmLTestSource keyboards get the get_key_list() compared to the `key2.kmap` table.
- fix a test case to not leak!

also

 - move some test utils to statics
 - print warning on unhanded @-commands
 - move kmxplus loading up into the LdmlTestSource parent class, so that all ldml test sources can access the KMX+ binary format for inspection and verification.
 - update some test utility functions - dump out a key_event as a string, etc.

Fixes: #12298

(tests from https://github.com/keymanapp/keyman/pull/12281 )

# User Testing

Please paste in the following LDML keyboard source for testing:

```xml
<?xml version="1.0"?>
<keyboard3 xmlns="https://schemas.unicode.org/cldr/45/keyboard3" locale="arc" conformsTo="45">
  <info author="Marc Durdin" name="Test 'other' modifier"/>
  <version number="1.0.0"/>
  <keys>
    <import base="cldr" path="45/keys-Zyyy-punctuation.xml"/>
    <import base="cldr" path="45/keys-Zyyy-currency.xml"/>
  	<!-- switch keys -->

    <key id="a" output="🅐" />
    <key id="b" output="🅑" />
    <key id="c" output="🅒" />
    <key id="d" output="🅓" />
    <key id="e" output="🅔" />
  </keys>
  <layers formId="us">
    <layer modifiers="none">
      <row keys="grave 1 2 3 4 5 6 7 8 9 0 hyphen equal"/>
      <row keys="a"/>
      <row keys="gap"/>
      <row keys="gap"/>
      <row keys="space"/>
    </layer>
    <layer modifiers="shift">
      <row keys="tilde bang at hash dollar percent caret amp asterisk open-paren close-paren underscore plus"/>
      <row keys="b"/>
      <row keys="gap"/>
      <row keys="gap"/>
      <row keys="space"/>
    </layer>
    <layer modifiers="ctrl">
      <row keys="grave 1 2 3 4 5 6 7 8 9 0 hyphen equal"/>
      <row keys="c"/>
      <row keys="gap"/>
      <row keys="gap"/>
      <row keys="space"/>
    </layer>
    <layer modifiers="other">
      <row keys="grave 1 2 3 4 5 6 7 8 9 0 hyphen equal"/>
      <row keys="d e"/>
      <row keys="gap"/>
      <row keys="gap"/>
      <row keys="space"/>
    </layer>
  </layers>
</keyboard3>
```

Create a basic LDML keyboard, and paste in the source above. Compile the keyboard and its package. Use this keyboard in all the subsequent tests.

For all tests, pass the test if:

* <kbd>q</kbd> emits 🅐
* <kbd>shift</kbd>+<kbd>q</kbd> emits 🅑
* <kbd>ctrl</kbd>+<kbd>q</kbd> emits 🅒
* <kbd>alt</kbd>+<kbd>q</kbd> emits 🅓
* <kbd>alt</kbd>+<kbd>w</kbd> emits 🅔
* the following key combinations have no output:
  * <kbd>alt</kbd>+<kbd>e</kbd>
  * <kbd>ctrl</kbd>+<kbd>e</kbd>
  * <kbd>shift</kbd>+<kbd>e</kbd>
  * <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>e</kbd>
  * <kbd>ctrl</kbd>+<kbd>alt</kbd>+<kbd>e</kbd>
  * <kbd>e</kbd>

* **TEST_DEBUGGER:** Test the keyboard in the Keyman Developer Debugger
* **TEST_WINDOWS:** Install the keyboard in Keyman for Windows, and test the keyboard in Notepad
* **TEST_MAC:** Install the keyboard in Keyman for macOS, and test the keyboard in TextEdit
* **TEST_LINUX:** Install the keyboard in Keyman for Linux, and test the keyboard in a text editor

srl295 commented on core/tests/unit/ldml/keyboards/k_004_tinyshift.xml:
```diff
@@ -4,6 +4,7 @@
 
 @@keys: [SHIFT K_BKQUOTE][K_1][K_BKQUOTE]
 @@expected: \u0037\u1790\u17B6\u0127
+@@keylist: [SHIFT K_BKQUOTE][SHIFT K_1][K_BKQUOTE][K_1]
```
Just picked this because it's a smaller keyboard.

srl295 commented:
@rc-swag @mcdurdin would this need user testing?

also note that gap keys / unassigned keys are NOT included in the list.

mcdurdin commented:
> @rc-swag @mcdurdin would this need user testing?

Yes, I think it should have testing (all desktop platforms)

> also note that gap keys / unassigned keys are NOT included in the list.

They probably need to be, otherwise they will fall back to default output (think Shift modifier for example). The rule we decided on for LDML keyboards was that if a layer is defined at all for a given modifier set, it is defined for all keys on that modifier set (even if it is no-op).

srl295 commented:
>> @rc-swag @mcdurdin would this need user testing?

> Yes, I think it should have testing (all desktop platforms)

- [ ] OK. Will ponder the plan.

>> also note that gap keys / unassigned keys are NOT included in the list.

> They probably need to be, otherwise they will fall back to default output (think Shift modifier for example). The rule we decided on for LDML keyboards was that if a layer is defined at all for a given modifier set, it is defined for all keys on that modifier set (even if it is no-op).

- [ ] OK. So the set will be more fixed, more based on the modifier set than the exact key layout.  I can probably just generate the list then.

rc-swag commented:
> > > @rc-swag @mcdurdin would this need user testing?
> 
> > Yes, I think it should have testing (all desktop platforms)
> 
> * [ ]  OK. Will ponder the plan.

Following the test the found the issue in the first place will exercise the code. Not sure for the other platforms if there is a ready made user test.

srl295 commented:
great, yes bringing in the test case from https://github.com/keymanapp/keyman/issues/12298#issue-2488657340 makes sense.

i think i could do this with a set of the modifier combinations, and then just expand on the 'vkey map' X modifier combinations.

srl295 commented:
@mcdurdin @rc-swag I think this is ready for some code review to start.  I'm working on the user testing story now.

mcdurdin commented:
> @mcdurdin @rc-swag do the user tests look OK? copied them from #12281
> 
> I was not able to reproduce a change before/after this using MSVC with testhost.

Yes, user tests look good to me

rc-swag commented:
> I was not able to reproduce a change before/after this using MSVC with testhost.

See my comment in the PR Draft https://github.com/keymanapp/keyman/issues/12298#issuecomment-2502889262
I am not seeing a change yet either he ALT key is still not being processed in the Keyman Hook.

dinakaranr commented:
# Test Results
I tested this issue with the attached "Keyman Developer 18.0.146-alpha-test-12644" build on Windows 10.  I'm sharing my observation here.
Prerequisite steps:
1. Installed the Keyman developer 18.0.146-alpha-test-12644.exe build from Team City.
2. Open the keyman developer application.
3. Create a new "LDML Keyboard" project. (While creating a name, language)
4. Navigate to the "Keyboards" tab.
5. Click the *.XML link to open the new tab for XML content.
6. Paste the source given in this task and save it.
7. Navigate to the "Keyboards" tab.
8. Click the "Options" button, and then select "Build.".
9. Navigate to the "Packaging" tab, and then click the "Build All" button.
I observed that the "Built successfully" message appears.

* **TEST_DEBUGGER (Passed):**
Keyman Debugger:
1. Installed the Keyman 18.0.146-alpha version and then started the configuration.
2. On Keyman Developer, navigate to the "Keyboard" tab and then click the *.kmn link button to navigate to the build tab.
3. Click the "Start Debugging" button or press F5 key.
4. Debugger mode was opened.
5. An empty pane appears and then click in the box.
6. Press the below keys and observe the below option.
q emits "A" appeared. It's correct
shift+q emits "B" appeared. It's correct
ctrl+q emits "C" appeared. It's correct
alt+q emits "D" appeared. It's correct
alt+w emits "E" appeared. It's correct
The below key combination does not give any outputs: It correct
alt+e, ctrl+e, shift+e, ctrl+shift+e, ctrl+alt+e, e
"?" box appeared when pressing the "q" but the character grid showed the right values Hence, I passed this scenario. Thanks. Please refer to the screenshot below.
<img width="300" src="https://github.com/user-attachments/assets/8826909e-479d-4837-9a9d-a7f2efa091e6">

* **TEST_WINDOWS (passed):** 
Window 10 OS:
1. I installed the Keyman 18.0.146-alpha-test-12644 version and then started the configuration.
2. On Keyman developer: Navigate to the "Packaging" tab and then click the *.kmx link button to install the keyboard in Windows 10.
3. Select the LDML keyboard that was installed through "Keyman Developer"
4. Open the notepad. Press the below keys and observe the below options.
5. q emits "A" and appears. It's correct.
shift+q emits "B" appeared. It's correct
ctrl+q emits "C" appeared. It's correct
alt+q emits 🅓 no results
alt+w emits 🅔 no results
The below key combination does not give any outputs: It correct
alt+e, ctrl+e, shift+e, ctrl+shift+e, ctrl+alt+e, e
This behavior already has an issue in Github and the PR number is #12298 hence, I passed this test case.

* **TEST_MAC (Passed):**
MAC OS:
1. Installed the Keyman keyboard package version 18.0.146-alpha-local
2. Open the configuration dialog.
3. Installed the LDML keyboard that was created in Keyman Developer (by clicking the *.kmp file).
4. Open the "Text Editor" application.
5. Select the LDML keyboard in the toolbar.
6. Press the below keys and observe the below option
q emits "A" appeared. It's correct
shift+q emits "B" appeared. It's correct
ctrl+q emits "C" appeared. It's correct
alt+q emits "D" appeared. It's correct
alt+w emits "E" appeared. It's correct
The below key combination does not give any outputs: It correct
alt+e, ctrl+e, shift+e, ctrl+shift+e, ctrl+alt+e, e

* **TEST_LINUX (Passed):**
Ubuntu Focal OS:
1. Installed the Keyman keyboard package version 18.0.146-1~PR-12644-4048.1+focal1
2. Open the configuration dialog.
3. Installed the LDML keyboard that was created in Keyman Developer (by selecting the *.kmp file)
4. Open the "Text Editor" application.
5. Select the LDML keyboard in the toolbar.
6. Press the below keys and observe the below options
q emits "A" appeared. It's correct
shift+q emits "B" appeared. It's correct
ctrl+q emits "C" appeared. It's correct
alt+q emits "D" appeared. It's Correct
alt+w emits "E" appeared. It's Correct
The below key combination does not give any outputs: It correct
alt+e, ctrl+e, shift+e, ctrl+shift+e, ctrl+alt+e, e

srl295 commented:
Thanks @dinakaranr

srl295 commented:
I updated the k_004 test case with:

```xml
    <layer id="base" modifiers="none">
      <row keys="hmaqtugha that" /> <!-- number row -->
    </layer>
    <layer id="shift" modifiers="shift">
      <row keys="seven eee" /> <!-- number row -->
    </layer>
    <layer id="control" modifiers="ctrl">
      <row keys="that gap" /> <!-- number row -->
    </layer>
    <layer id="catchall" modifiers="other">
      <row keys="eee gap" /> <!-- number row -->
    </layer>
```

I can confirm that I'm seeing modifiers in the `get_key_list()` including:

- LDML_KEYS_MOD_NONE 0x0
- LDML_KEYS_MOD_SHIFT 0x10
- LDML_KEYS_MOD_OTHER 0x10000
- LDML_KEYS_MOD_CTRL 0x20

mcdurdin reviewed (approved):
> * note that ALL keys are included, not just assigned keys.

What does ALL keys mean in this context? Do you mean all key+modifier combinations for which there is at least one key on a given layer? Is no-modifier always included? Does 'ALL' include 102nd key (right of left shift) if the form is ANSI (or extra ABNT2 key?)? I'd just like to make sure we are all on the same page here :grin:

srl295 commented:
> > * note that ALL keys are included, not just assigned keys.
> 
> What does ALL keys mean in this context?

all keys in the <form>

> Do you mean all key+modifier combinations for which there is at least one key on a given layer?

yes, but

> Is no-modifier always included?

that i would have to check, perhaps a keyboard could be defined that didn't have a no-modifier layer.

> Does 'ALL' include 102nd key (right of left shift) if the form is ANSI (or extra ABNT2 key?)? I'd just like to make sure we are all on the same page here 😁

only includes keys on the form. so no, not all possible keys.

But as I noted it uses the modifier OTHER and does not expand it to the actual modifiers.

rc-swag commented:
> But as I noted it uses the modifier OTHER and does not expand it to the actual modifiers.

I thought you may have already done the expansion before the add keys part.  In the issue for this PR I commented on this I stated that. https://github.com/keymanapp/keyman/issues/12298#issuecomment-2453953772
Implementation detail:
Of interest with this example also is that the ALT key is in a keyboard rule via the the LDML rule other for modifiers. The get_key_list() will obviously have to do the logical expansion of the rule.

The windows engine to my knowledge has no concept of OTHER.

mcdurdin commented:
> But as I noted it uses the modifier OTHER and does not expand it to the actual modifiers.

The `OTHER` modifier is an internal detail of the LDML implementation and must not leak out into Core.

srl295 commented:
@rc-swag  you did state it in the issue. Will fix it.

>> Of interest with this example also is that the ALT key is in a keyboard rule via the the LDML rule other for modifiers. The get_key_list() will obviously have to do the logical expansion of the rule.

I will note though, that even if I explicitly set the layer to be 'alt' modifier (and not other) I don't think I'm seeing the correct behavior (i.e. alt-q, alt-w) locally. 

I will work on the expansion, but as noted let's try to debug next wk

rc-swag commented:
@srl295 Ok I built the latest commit of this and tested a build of the keyboard in the user testing section. The alt (including RALT) key behaves correctly and ouputs a `🅓`. I also replace `other` with `alt` and  that also worked.  To see have preserved keys printed into the log I commented #define _DEBUG_LOGKEY in ln.38 of keys.cpp. That allow me to see that it appeared to be working as expected.


```
CKMTipTextService::_PreserveAltKeys: Preserved key 162: K_Q 0, 0
CKMTipTextService::_PreserveAltKeys: Preserved key 163: K_Q 4, 0
CKMTipTextService::_PreserveAltKeys: Preserved key 164: K_Q 2, 0
CKMTipTextService::_PreserveAltKeys: Preserved key 165: K_Q 10, 
CKMTipTextService::_PreserveAltKeys: Preserved key 166: K_Q 40, 
CKMTipTextService::_PreserveAltKeys: Preserved key 167: K_Q 8, 0
```

srl295 commented:
> @srl295 Ok I built the latest commit of this and tested a build of the keyboard in the user testing section. The alt (including RALT) key behaves correctly and ouputs a `🅓`. As seperate keyboard test I also replaced `other` with `alt` and that also worked. To see have preserved keys printed into the log I commented #define _DEBUG_LOGKEY in ln.38 of keys.cpp. That allowed me to see that it appeared to be working as expected.
> 
> ```
> CKMTipTextService::_PreserveAltKeys: Preserved key 162: K_Q 0, 0
> CKMTipTextService::_PreserveAltKeys: Preserved key 163: K_Q 4, 0
> CKMTipTextService::_PreserveAltKeys: Preserved key 164: K_Q 2, 0
> CKMTipTextService::_PreserveAltKeys: Preserved key 165: K_Q 10, 
> CKMTipTextService::_PreserveAltKeys: Preserved key 166: K_Q 40, 
> CKMTipTextService::_PreserveAltKeys: Preserved key 167: K_Q 8, 0
> ```

I missed this but this is great news.

I've updated the code to do the combinatoric explosion as discussed. Passes tests (which do a spot check of the expansions.)

Think this may be ready to go.

srl295 commented:
> Thanks @dinakaranr - actually
> 
> > (Windows 10)
> > alt+q emits 🅓 no results
> > alt+w emits 🅔 no results
> > The below key combination does not give any outputs: It correct
> > alt+e, ctrl+e, shift+e, ctrl+shift+e, ctrl+alt+e, e
> > This behavior already has an issue in Github and the PR number is #12298 hence, I passed this test case.
> 
> Actually, you are right that it is already an issue, however, this PR is intended to fix the issue - therefore, please fail the test case.

@rc-swag should this test case now pass? ☝️

rc-swag reviewed (approved):
lgtm

rc-swag commented:
> @rc-swag should this test case now pass? ☝️

It should. 
@keymanapp-test-bot retest all

dinakaranr commented:
# Test Results

* **TEST_WINDOWS (Passed):**
Window 10 OS:
1. I installed the Keyman 18.0.156-alpha-test-12644 version and then started the configuration.
2. On Keyman developer: Navigate to the "Packaging" tab and then click the *.kmx link button to install the keyboard in Windows 10.
3. Select the LDML keyboard that was installed through "Keyman Developer"
4. Open the notepad. Press the keys below and observe the options below.
5. q emits "A" and appears. It's correct.
6. shift+q emits "B" appeared. It's correct
7. ctrl+q emits "C" appeared. It's correct
8. alt+q emits "D"appeared. It's correct.
9. alt+w emits "E" appeared. It's correct.
10. The below key combination does not give any outputs: It's correct

alt+e, ctrl+e, shift+e, ctrl+shift+e, ctrl+alt+e, e
The D and E keys appeared on the notepad. It works great for me. Thank you.
Please take a look at the screenshot
<img width="300" src="https://github.com/user-attachments/assets/f884b87e-6afa-4646-b697-31f04548acc2">

mcdurdin commented:
Can we get a cherry-pick of this to 17.0 please?
